docs/proposals/petset.md

```diff
@@ -0,0 +1,533 @@
+
+
+```
+type PetSetSpec struct {
+    // Number of instances in the set
+    Replicas *int
+
+    // The template to use for the pod in the set
+    Template *PodTemplateSpec
+}
+
+// Proposal 1a: Use the name of the pod as the identity, use downward API for getting identity
+// Downside: all fields that want to leverage the identity need equivalent valueFrom semantics
+{
+    "spec": {
+        "template": {
+            "metadata": {"generateName": "pod-name"}, // assume identity is automatically added to pod name
+            "spec": {
+                "containers": [{
+                    "env": [{
+                        "name": "IDENTITY",
+                        "valueFrom": {"fieldRef": {"fieldPath": "metadata.name"}
+                        }
+                    }]
+                }]
+            }
+        }
+    }
+}
+
+// Proposal 1b: Add an annotation to each pod when the controller creates it
+// Downside: all fields that want to leverage the identity need equivalent valueFrom semantics
+{
+    "spec": {
+        "template": {
+            "spec": {
+                "containers": [{
+                    "env": [{
+                        "name": "IDENTITY",
+                        "valueFrom": {"fieldRef": {"fieldPath": "metadata.annotations[alpha.kubernetes.io/identity]"}}
+                    }]
+                }]
+            }
+        }
+    }
+}
+
+// Proposal 1c: Add a new field to pod spec that is write-once/read-many - identity
+// Downside: all fields that want to leverage the identity need equivalent valueFrom semantics
+{
+    "spec": {
+        "template": {
+            "spec": {
+                "identity": "", // set on creation
+                "containers": [{
+                    "env": [{
+                        "name": "IDENTITY",
+                        "valueFrom": {"fieldRef": {"fieldPath": "spec.identity"}}
+                    }]
+                }]
+            }
+        }
+    }
+}
+
+// Proposal 1d: support template transformation of a field path
+// Downside: possibly too flexible, can conflict with existing env resolution semantics, bypasses downward API
+{
+    "spec": {
+        "template": {
+            "metadata": {"name": "pod-name-$(IDENTITY)"},
+            "spec": {
+                "containers": [{
+                    "env": [{
+                        "name": "IDENTITY",
+                        "value": "$(IDENTITY)"
+                    }]
+                }]
+            }
+        }
+    }
+}
+
+// Proposal 1e: specify merge-patch semantics with some variable replacement
+// Downside: merge patch is inflexible to structural template edits
+{
+    "spec": {
+        "template": {
+            "metadata": {},
+            "spec": {
+                "containers": [{
+                    "env": [{
+                        "name": "IDENTITY",
+                    }]
+                }]
+            }
+        },
+        "templateTransform": [
```

Comment:
I agree... except that all the others are non-descriptive transforms that
don't require excessive API additions.  The general purpose processing,
while what we implemented in templates, has problems transforming
versions.  I wanted to at least explore something that is pretty clear
(take the template, apply these changes).  To a naive client, the fields
that will be modified are obvious.

On Thu, Dec 3, 2015 at 8:07 PM, Brian Grant <notifications@github.com>
wrote:

> In docs/proposals/petset.md
> <https://github.com/kubernetes/kubernetes/pull/18016#discussion_r46637028>
> :
>
> > +
> > +// Proposal 1e: specify merge-patch semantics with some variable replacement
> > +// Downside: merge patch is inflexible to structural template edits
> > +{
> > +    "spec": {
> > +        "template": {
> > +            "metadata": {},
> > +            "spec": {
> > +                "containers": [{
> > +                    "env": [{
> > +                        "name": "IDENTITY",
> > +                    }]
> > +                }]
> > +            }
> > +        },
> > +        "templateTransform": [
>
> This looks very complicated to me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/18016/files#r46637028>.
>